### PR TITLE
change FindServiceNetwork algo, use name and id match

### PR DESCRIPTION
**Pull request by mikhail-aws**
Note:

Address bug for RAM sharing scenario with ServiceNetwork. Currently FindServiceNetwork will not find shared SNs, but if find will fail on fetching Tags. It's not possible to fetch tags for external resources.

Important changes in vpclattice.go

New algorithm will do:
- List all SN
- Filter by Name match
- Filter by ID match. I use ID as fallback strategy. If there multiple SN with same name (which is possible with RAM sharing, but not in same account). Operator can create GW with SN-ID as a name to disambiguate naming.

Added unit tests.
Manually tested with shared SN scenarios.

New error log example:
```
reconcile error	{"name": "share-to-mhlbrz", "message": "multiple SN found by name, cannot disambiguate: [name=share-to-mhlbrz, id=sn-09ee4f73b2985af2b name=share-to-mhlbrz, id=sn-08542bb99cb2f3bae]"}
```

Related #565

### CODE REVIEW

File: pkg/aws/services/vpclattice.go [hunk 1]
```diff
@@ -313,60 +318,78 @@ func (d *defaultLattice) ListServiceNetworkServiceAssociationsAsList(ctx context
...
 	}
-	if snMatch == nil {
-		return nil, NewNotFoundError("Service network", name)
+
+	// try to fetch tags only if SN in the same aws account with controller's config
+	tags := Tags{}
+	if snArn.AccountID == d.ownAccount || d.ownAccount == "" {
+		tagsInput := vpclattice.ListTagsForResourceInput{ResourceArn: snMatch.Arn}
+		tagsOutput, err := d.ListTagsForResourceWithContext(ctx, &tagsInput)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (mikhail-aws) [reply pkg/aws/services/vpclattice.go:378]: It's true that we are not using tags for SN. But we might need them for cross-cluster same aws account sharing. For example for read-only access. I'm ok to remove them, but I have sense that we will need them very soon once we dig more into cross-cluster use-cases.

File: pkg/aws/services/vpclattice.go [hunk 2]
```diff
@@ -104,17 +105,17 @@ type Lattice interface {
 	ListTargetsAsList(ctx context.Context, input *vpclattice.ListTargetsInput) ([]*vpclattice.TargetSummary, error)
 	ListServiceNetworkVpcAssociationsAsList(ctx context.Context, input *vpclattice.ListServiceNetworkVpcAssociationsInput) ([]*vpclattice.ServiceNetworkVpcAssociationSummary, error)
 	ListServiceNetworkServiceAssociationsAsList(ctx context.Context, input *vpclattice.ListServiceNetworkServiceAssociationsInput) ([]*vpclattice.ServiceNetworkServiceAssociationSummary, error)
-	FindServiceNetwork(ctx context.Context, name string, accountId string) (*ServiceNetworkInfo, error)
+	FindServiceNetwork(ctx context.Context, name string) (*ServiceNetworkInfo, error)
```
Comment (zijun726911) [inline pkg/aws/services/vpclattice.go:108]: Need to change name to  `nameOrId string` in interface?

File: pkg/aws/services/vpclattice.go [hunk 3]
```diff
@@ -313,60 +318,78 @@ func (d *defaultLattice) ListServiceNetworkServiceAssociationsAsList(ctx context
...
+	// match. Ideally name match should work just fine, but in desperate
+	// scenario of shared SN when naming collision happens using ID can be
+	// an option
+	var snMatch *vpclattice.ServiceNetworkSummary
+	nameMatch := utils.SliceFilter(allSn, func(snSum *vpclattice.ServiceNetworkSummary) bool {
+		return aws.StringValue(snSum.Name) == nameOrId
+	})
+	idMatch := utils.SliceFilter(allSn, func(snSum *vpclattice.ServiceNetworkSummary) bool {
+		return aws.StringValue(snSum.Id) == nameOrId
 	})
```
Comment (zijun726911) [inline pkg/aws/services/vpclattice.go:351]: For the discussion in slack, filter by name then id make sense. We could also add another filter by ARN, i.e. change function name to 
`FindServiceNetwork(ctx context.Context, nameOrIdOrArn string)` and add a `arnMatch=`

File: pkg/aws/services/vpclattice.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (zijun726911) [reply pkg/aws/services/vpclattice.go:378]: > we might need them for cross-cluster same aws account sharing

I am still not quite sure what same account cross-cluster sharing need tags. How's the SN read-only access looks like?

File: pkg/aws/services/vpclattice.go (hunk 3 above)
Comment (mikhail-aws) [reply pkg/aws/services/vpclattice.go:351]: I would drop ID and use ARN only

File: pkg/aws/services/vpclattice.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikhail-aws) [reply pkg/aws/services/vpclattice.go:378]: It's really hard to make SN fit into Gateway spec and include lattice features.

File: pkg/aws/services/vpclattice.go (hunk 3 above)
Comment (mikhail-aws) [reply pkg/aws/services/vpclattice.go:336]: ARN is not a valid name for Gateway, "/" if not allowed

**Review by zijun726911 — APPROVED** [review]

**Pull request merged**